src/core/ComboBox/ComboBox.tsx

```diff
@@ -180,189 +174,274 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
-          break;
-        default:
-          if (!isOpen) {
-            setIsOpen(true); // reopen menu if closed when typing
-          }
-          break;
-      }
     },
-    [focusedIndex, isOpen, options, getOptionId],
+    [focusedIndex, itemRenderer, memoizedItems, selectedIndex],
   );
 
-  const menuItems = React.useMemo(() => {
-    if (filteredOptions.length === 0) {
-      return [
-        <MenuExtraContent key={0}>
-          <Text isMuted>{emptyStateMessage}</Text>
-        </MenuExtraContent>,
-      ];
+  return (
+    <ComboBoxRefsContext.Provider
+      value={{ inputRef, menuRef, toggleButtonRef, optionsRef }}
+    >
+      <ComboBoxActionContext.Provider value={dispatch}>
+        <ComboBoxStateContext.Provider
+          value={{ id, minWidth, isOpen, focusedIndex }}
+        >
+          {children ?? (
+            <>
+              <ComboBoxInputContainer
+                disabled={inputProps?.disabled}
+                endIcon={
+                  <ComboBoxEndIcon
+                    className={cx({
+                      'iui-actionable': !inputProps?.disabled,
+                      'iui-disabled': inputProps?.disabled,
+                      'iui-open': isOpen,
+                    })}
+                  />
+                }
+                {...rest}
+              >
+                <ComboBoxInput
+                  value={inputValue}
+                  {...inputProps}
+                  onChange={handleOnInput}
+                />
+              </ComboBoxInputContainer>
+              <ComboBoxPopover {...dropdownMenuProps}>
+                <ComboBoxMenu>
+                  {filteredOptions.length > 0 ? (
+                    filteredOptions.map(getMenuItem)
+                  ) : (
+                    <MenuExtraContent>
+                      <Text isMuted>{emptyStateMessage}</Text>
+                    </MenuExtraContent>
+                  )}
+                </ComboBoxMenu>
+              </ComboBoxPopover>
+            </>
+          )}
+        </ComboBoxStateContext.Provider>
+      </ComboBoxActionContext.Provider>
+    </ComboBoxRefsContext.Provider>
+  );
+};
+
+type ComboBoxAction = 'open' | 'close' | 'toggle' | 'select' | 'focus';
+
+const comboBoxReducer = (
+  state: { isOpen: boolean; selectedIndex: number; focusedIndex: number },
+  [type, value]: [ComboBoxAction] | [ComboBoxAction, number | undefined],
+) => {
+  switch (type) {
+    case 'open': {
+      return { ...state, isOpen: true };
     }
-    return filteredOptions.map((option) => {
-      const index = options.findIndex(({ value }) => option.value === value);
-      if (index < 0) {
-        return <></>;
-      }
+    case 'close': {
+      return { ...state, isOpen: false };
+    }
+    case 'toggle': {
+      return { ...state, isOpen: !state.isOpen };
+    }
+    case 'select': {
+      return {
+        ...state,
+        isOpen: false,
+        selectedIndex: value ?? state.focusedIndex ?? state.selectedIndex,
```

Comment:
How can selected index be focused index?

Reply:
it selects the currently focused index if value is not passed in the dispatch

Reply:
Does it make sense to select focused item? I would say it should be the other way arround.

Reply:
I have both cases.
![image](https://user-images.githubusercontent.com/9084735/165173138-9512afc5-a55d-4791-a293-ee034e3990c1.png)